File: tests/basic_test/publisher.py
import sys
import time
from datetime import timedelta

from pyopendds import \
    init_opendds, DomainParticipant, StatusKind, PyOpenDDS_Error
from pybasic.basic import *

if __name__ == "__main__":
    try:
        # Initialize OpenDDS and Create DDS Entities
        init_opendds(opendds_debug_level=1)
        time.sleep(1)
        domain = DomainParticipant(34)
        time.sleep(1)
        topic = domain.create_topic('Readings', Reading)
        time.sleep(1)
        publisher = domain.create_publisher()
        time.sleep(1)
        writer = publisher.create_datawriter(topic)
        time.sleep(1)

        # Wait for Subscriber to Connect
        print('Waiting for Subscriber...')
        writer.wait_for(StatusKind.PUBLICATION_MATCHED, timedelta(seconds=60))
        print('Found subscriber!')

        write_sample_speed = Reading()
        s1 = Sample()
        s1.value = 1
        s1.where = "toto1"

        s2 = Sample()
        s2.value = 2
        s2.where = "toto2"

        write_sample_accel = Reading()
        write_sample_dist = Reading()

        while True:
            time.sleep(1)
            write_sample_speed.kind = ReadingKind.speed
            write_sample_speed.value = 123
            write_sample_speed.where = "somewhere"
            write_sample_speed.sampleSeq = seqSample([s1, s2])
            # Read and Print Sample
            rc = writer.write(write_sample_speed)
            print(rc)

            time.sleep(1)
            write_sample_accel.kind = ReadingKind.acceleration
            write_sample_accel.value = 2
            write_sample_accel.where = "everywhere"
            write_sample_accel.sampleSeq = seqSample([s1, s2])
            # Read and Print Sample
            rc = writer.write(write_sample_accel)
            print(rc)

            time.sleep(1)
            write_sample_dist.kind = ReadingKind.distance
            write_sample_dist.value = 543
            write_sample_dist.where = "anywhere"
            write_sample_dist.sampleSeq = seqSample([s1, s2])
            # Read and Print Sample
            rc = writer.write(write_sample_dist)
            print(rc)
            print('Done!')

    except PyOpenDDS_Error as e:
        sys.exit(e)

File: tests/basic_test/publisher.cpp
#include <basicTypeSupportImpl.h>

#include <dds/DdsDcpsInfrastructureC.h>
#include <dds/DdsDcpsPublicationC.h>
#include <dds/DCPS/Marked_Default_Qos.h>
#include <dds/DCPS/Service_Participant.h>
#include <dds/DCPS/WaitSet.h>
#include <dds/DCPS/DCPS_Utils.h>

#include <iostream>

using OpenDDS::DCPS::retcode_to_string;

int main(int argc, char* argv[])
{
  try {
    // Init OpenDDS
    TheServiceParticipant->default_configuration_file("rtps.ini");
    DDS::DomainParticipantFactory_var opendds = TheParticipantFactoryWithArgs(argc, argv);

    DDS::DomainParticipantQos part_qos;
    opendds->get_default_participant_qos(part_qos);
    DDS::DomainParticipant_var participant =
      opendds->create_participant(34, part_qos, 0, OpenDDS::DCPS::DEFAULT_STATUS_MASK);
    if (!participant) {
      std::cerr << "Error: Failed to create participant" << std::endl;
      return 1;
    }

    basic::ReadingTypeSupport_var ts = new basic::ReadingTypeSupportImpl();
    DDS::ReturnCode_t rc = ts->register_type(participant.in(), "");
    if (rc != DDS::RETCODE_OK) {
      std::cerr << "Error: Failed to register type: " << retcode_to_string(rc) << std::endl;
      return 1;
    }

    CORBA::String_var type_name = ts->get_type_name();
    DDS::Topic_var topic = participant->create_topic(
      "Readings", type_name.in(), TOPIC_QOS_DEFAULT, 0, OpenDDS::DCPS::DEFAULT_STATUS_MASK);
    if (!topic) {
      std::cerr << "Error: Failed to create topic" << std::endl;
      return 1;
    }

    DDS::Publisher_var publisher =
      participant->create_publisher(PUBLISHER_QOS_DEFAULT, 0, OpenDDS::DCPS::DEFAULT_STATUS_MASK);
    if (!publisher) {
      std::cerr << "Error: Failed to create publisher" << std::endl;
      return 1;
    }

    DDS::DataWriterQos qos;
    publisher->get_default_datawriter_qos(qos);
    DDS::DataWriter_var writer =
      publisher->create_datawriter(topic.in(), qos, 0, OpenDDS::DCPS::DEFAULT_STATUS_MASK);
    if (!writer) {
      std::cerr << "Error: Failed to create writer" << std::endl;
      return 1;
    }

    // Wait for Subscriber
    std::cout << "Wating for Subscriber..." << std::endl;
    DDS::StatusCondition_var sc = writer->get_statuscondition();
    sc->set_enabled_statuses(DDS::PUBLICATION_MATCHED_STATUS);
    DDS::WaitSet_var ws = new DDS::WaitSet;
    ws->attach_condition(sc);
    const DDS::Duration_t max_wait = {10, 0};
    DDS::PublicationMatchedStatus status = {0, 0, 0, 0, 0};
    while (status.current_count < 1) {
      DDS::ConditionSeq active;
      if (ws->wait(active, max_wait) != DDS::RETCODE_OK) {
        std::cerr << "Error: Timedout waiting for subscriber" << std::endl;
        return 1;
      }
      if (writer->get_publication_matched_status(status) != DDS::RETCODE_OK) {
        std::cerr << "Error: Failed to get pub matched status" << std::endl;
        return 1;
      }
    }
    ws->detach_condition(sc);
    std::cout << "Found Subscriber..." << std::endl;

    ACE_OS::sleep(1);

    // Write Sample
    basic::ReadingDataWriter_var reading_writer = basic::ReadingDataWriter::_narrow(writer);
    basic::Reading reading;
    reading.kind
#ifdef CPP11_IDL
      () = basic::ReadingKind::speed;
#else
      = basic::speed;
#endif
    reading.value
#ifdef CPP11_IDL
      ()
#endif
      = -200;
    reading.where
#ifdef CPP11_IDL
      ()
#endif
      = "Somewhere";
    rc = reading_writer->write(reading, DDS::HANDLE_NIL);
    if (rc != DDS::RETCODE_OK) {
      std::cerr << "Error: Failed to write: " << retcode_to_string(rc) << std::endl;
      return 1;
    }

    ACE_OS::sleep(1);

    // Cleanup
    participant->delete_contained_entities();
    opendds->delete_participant(participant.in());
    TheServiceParticipant->shutdown();

  } catch (const CORBA::Exception& e) {
    e._tao_print_exception("Exception caught in main():");
    return 1;
  }

  return 0;
}
